### PR TITLE
Add ability to sign metadata. (Improved)

### PR DESCRIPTION
**Pull request by pitbulk**
Related to #207.
Improved:
* Test: Added signature validation
* prettyprint optional on metadata generator in order to avoid signature validation problems.
* signature position fixed. Now is placed at the top,

Fix algorithm calculator. This method was wrong

### CODE REVIEW

**Comment by pitbulk** [comment]
@Lordnibbler, @luisvm what do you think about this PR?

File: test/metadata_test.rb [hunk 1]
```diff
@@ -5,86 +5,117 @@
...
+
+    describe "when the settings indicate to sign (embedded) the metadata" do
+      before do
+        settings.security[:metadata_signed] = true
+        settings.certificate = ruby_saml_cert_text
+        settings.private_key = ruby_saml_key_text
+      end
+
+      it "creates a signed metadata" do
+        assert_match %r[<ds:SignatureValue>\s*([a-zA-Z0-9/+=]+)\s*</ds:SignatureValue>]m, xml_text
```
Comment (Umofomia) [inline test/metadata_test.rb]: FYI, I originally added the two `\s*` in the regular expression and the `m` at the end of it to take into account the fact that the metadata was being pretty-printed. Now that you took out the pretty-printing for the test, these should no longer be needed.

File: test/metadata_test.rb [hunk 2]
```diff
@@ -5,86 +5,117 @@
...
+      end
+
+      describe "when digest and signature methods are specified" do
+        before do
+          settings.security[:signature_method] = XMLSecurity::Document::RSA_SHA256
+          settings.security[:digest_method] = XMLSecurity::Document::SHA512
+        end
+
+        it "creates a signed metadata with specified digest and signature methods" do
+          assert_match %r[<ds:SignatureValue>\s*([a-zA-Z0-9/+=]+)\s*</ds:SignatureValue>]m, xml_text
```
Comment (Umofomia) [inline test/metadata_test.rb]: Same thing here.

File: test/metadata_test.rb (hunk 1 above)
Comment (pitbulk) [inline test/metadata_test.rb]: Fixed

File: test/metadata_test.rb [hunk 3]
```diff
@@ -5,86 +5,129 @@
...
     end
 
-    it "generates Service Provider Metadata with X509Certificate" do
-      @settings.security[:authn_requests_signed] = true
-      @settings.certificate = ruby_saml_cert_text
+    it "generates Pretty Print Service Provider Metadata" do
+      xml_text = OneLogin::RubySaml::Metadata.new.generate(settings, true)
+      # assert correct xml declaration
+      start = "<?xml version='1.0' encoding='UTF-8'?>\n<md:EntityDescriptor"
+      assert xml_text[0..start.length-1] == start
```
Comment (Lordnibbler) [inline test/metadata_test.rb]: probably good to use `assert_equal` here to be consistent

File: test/request_test.rb [hunk 4]
```diff
@@ -145,7 +145,7 @@ class RequestTest < Minitest::Test
       end
     end
 
-    describe "when the settings indicate to sign (embebed) the request" do
+    describe "when the settings indicate to sign (embedded) the request" do
```
Comment (Lordnibbler) [inline test/request_test.rb]:
```ruby
describe 'when the settings indicate to sign (embedded) request' do
```

File: test/slo_logoutresponse_test.rb [hunk 5]
```diff
@@ -67,7 +67,7 @@ class SloLogoutresponseTest < Minitest::Test
       assert_match /<samlp:StatusMessage>Custom Logout Message<\/samlp:StatusMessage>/, inflated
     end
 
-    describe "when the settings indicate to sign (embebed) the logout response" do
+    describe "when the settings indicate to sign (embedded) the logout response" do
```
Comment (Lordnibbler) [inline test/slo_logoutresponse_test.rb]:
```ruby
describe 'when the settings indicate to sign (embedded) logout response' do
```

**Comment by Lordnibbler** [comment]
comments aside looks good :+1:

**Pull request merged**